File: scripts/src/thunder-client.js
const request = require('request');

const Method = {
  POST:   { name: 'POST', expected: 201 },
  GET:    { name: 'GET', expected: 200 },
  PUT:    { name: 'PUT', expected: 200 },
  DELETE: { name: 'DELETE', expected: 200 },
  VERIFY: { name: 'VERIFY', expected: 200 },
  EMAIL:  { name: 'EMAIL', expected: 200 }
};

/**
 * The ThunderClient class provides methods to communicate to Thunder.
 */
class ThunderClient {
  /**
   * Constructs a new ThunderClient.
   * @constructor
   *
   * @param {string} endpoint - The base endpoint to connect to Thunder with.
   * @param {string} application - The name of the application key used for basic auth.
   * @param {string} secret - The value of the application secret used for basic auth.
   */
  constructor(endpoint, application, secret) {
    this.baseRequest = request.defaults({
      baseUrl: endpoint,
      auth:    { username: application, password: secret }
    });
  }

  /**
   * POST /users
   * Creates a new Thunder user.
   *
   * @param {object} user - The data of the user object to create.
   * @param {function} callback - The function to call when the method completes.
   */
  createUser(user, callback) {
    this.baseRequest.post({
      url:  '/users',
      body: user,
      json: true
    }, (err, res, body) => {
      if (err) return callback(err);

      return checkResponse(res, body, Method.POST, callback);
    });
  }

  /**
   * GET /users
   * Gets a Thunder user.
   *
   * @param {string} email - The email address of the user to get.
   * @param {string} password - The password of the user to get.
   * @param {function} callback - The function to call when the method completes.
   */
  getUser(email, password, callback) {
    this.baseRequest.get({
      url:     '/users',
      headers: { password: password },
      qs:      { email: email }
    }, (err, res, body) => {
      if (err) return callback(err);

      return checkResponse(res, body, Method.GET, callback);
    });
  }

  /**
   * PUT /users
   * Updates a Thunder user.
   *
   * @param {string} email - The existing email address of the user to update.
   * @param {string} password - The password of the user to update.
   * @param {object} user - The user object to PUT as an update.
   * @param {function} callback - The function to call when the method completes.
   */
  updateUser(email, password, user, callback) {
    this.baseRequest.put({
      url:     '/users',
      headers: { password: password },
      qs:      (email !== null && email !== '') ? { email: email } : {},
      body:    user,
      json:    true
    }, (err, res, body) => {
      if (err) return callback(err);

      return checkResponse(res, body, Method.PUT, callback);
    });
  }

  /**
   * DELETE /users
   * Deletes a Thunder user.
   *
   * @param {string} email - The email address of the user to delete.
   * @param {string} password - The password of the user to delete.
   * @param {function} callback - The function to call when the method completes.
   */
  deleteUser(email, password, callback) {
    this.baseRequest.delete({
      url:     '/users',
      headers: { password: password },
      qs:      { email: email }
    }, (err, res, body) => {
      if (err) return callback(err);

      return checkResponse(res, body, Method.DELETE, callback);
    });
  }

  /**
   * POST /verify
   * Creates and sends a new verification email.
   *
   * @param {string} email - The email address of the user to send the email to.
   * @param {string} password - The password of the user to send the email to.
   * @param {function} callback - The function to call when the method completes.
   */
  sendEmail(email, password, callback) {
    this.baseRequest.post({
      url:     '/verify',
      headers: { password: password },
      qs:      { email: email }
    }, (err, res, body) => {
      if (err) return callback(err);

      return checkResponse(res, body, Method.EMAIL, callback);
    });
  }

  /**
   * GET /verify
   * Verifies a user. Simulates the user clicking the URL in the email.
   *
   * @param {string} email - The email address of the user to verify.
   * @param {string} token - The verification token that should match the generated token.
   * @param {function} callback - The function to call when the method completes.
   */
  verifyUser(email, token, callback) {
    this.baseRequest.get({
      url: '/verify',
      qs:  { email: email, token: token }
    }, (err, res, body) => {
      if (err) return callback(err);

      return checkResponse(res, body, Method.VERIFY, callback);
    });
  }
}

/**
 * Checks if a Thunder response was successful.
 *
 * @param {object} res - The response object to verify.
 * @param {object} body - The reponse body.
 * @param {Method} method - The method that was called to produce the response.
 * @param {function} callback - The function to call when the method completes.
 * @return When the response check is complete.
 */
function checkResponse(res, body, method, callback) {
  if (res.statusCode !== method.expected) {
    return callback(
      new Error('The status code ' + res.statusCode
        + ' does not match expected ' + method.expected), body);
  }

  let result;

  try {
    result = JSON.parse(body);
  } catch (e) {
    result = body;
  }

  return callback(null, result);
}

module.exports = ThunderClient;

File: client/src/main/java/com/sanctionco/thunder/ThunderClient.java
package com.sanctionco.thunder;

import com.sanctionco.thunder.models.ResponseType;
import com.sanctionco.thunder.models.User;
import com.sanctionco.thunder.testing.ThunderClientFake;

import java.util.concurrent.CompletableFuture;

import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Query;

/**
 * Provides methods to interact with Thunder over HTTP requests. Construct an instance
 * of this class using {@link ThunderClientBuilder}.
 */
public interface ThunderClient {

  /**
   * Creates the user in the user database.
   *
   * @param user the user to create
   * @return a {@link CompletableFuture} that holds the created user after the request completes
   */
  @POST("users")
  CompletableFuture<User> postUser(@Body User user);

  /**
   * Updates the user in the user database.
   *
   * @param user the user to update with all fields updated
   * @param existingEmail the existing email of the user. This may be {@code null} if the user's
   *                      email is not being updated.
   * @param password the user's password
   * @return a {@link CompletableFuture} that holds the updated user after the request completes
   */
  @PUT("users")
  CompletableFuture<User> updateUser(@Body User user,
                                     @Query("email") String existingEmail,
                                     @Header("password") String password);

  /**
   * Updates the user in the user database. The user to update must have the same email address.
   * See {@link #updateUser(User, String, String)} to update a user's email address.
   *
   * @param user the user to update with all fields updated
   * @param password the user's password
   * @return a {@link CompletableFuture} that holds the updated user after the request completes
   */
  default CompletableFuture<User> updateUser(@Body User user,
                                             @Header("password") String password) {
    return updateUser(user, null, password);
  }

  /**
   * Gets the user with the given email address from the user database.
   *
   * @param email the user's email address
   * @param password the user's password
   * @return a {@link CompletableFuture} that holds the user after the request completes
   */
  @GET("users")
  CompletableFuture<User> getUser(@Query("email") String email,
                                  @Header("password") String password);

  /**
   * Deletes the user with the given email address from the user database.
   *
   * @param email the user's email address
   * @param password the user's password
   * @return a {@link CompletableFuture} that holds the deleted user after the request completes
   */
  @DELETE("users")
  CompletableFuture<User> deleteUser(@Query("email") String email,
                                     @Header("password") String password);

  /**
   * Sends a verification email to the user with the given email address.
   *
   * @param email the user's email address
   * @param password the user's password
   * @return a {@link CompletableFuture} that holds the user after the request completes
   */
  @POST("verify")
  CompletableFuture<User> sendVerificationEmail(@Query("email") String email,
                                                @Header("password") String password);

  /**
   * Verifies the user with the given email. This method will return the verified user in the
   * response. To get HTML in the response, see
   * {@link ThunderClient#verifyUser(String, String, ResponseType)}.
   *
   * @param email the user's email address
   * @param token the email verification token that was generated from sending an email
   * @return a {@link CompletableFuture} that holds the verified user after the request completes
   */
  @GET("verify")
  CompletableFuture<User> verifyUser(@Query("email") String email,
                                     @Query("token") String token);

  /**
   * Verifies the user with the given email. This method can return HTML in the response.
   * To get the verified user object in the response, see
   * {@link ThunderClient#verifyUser(String, String)}.
   *
   * @param email the user's email address
   * @param token the email verification token that was generated from sending an email
   * @param responseType the type of response to receive (HTML or JSON)
   * @return a {@link CompletableFuture} that holds the response after the request completes.
   *     The response will be an HTML string if responseType was set to HTML, or a JSON string if
   *     responseType was set to JSON.
   */
  @GET("verify")
  CompletableFuture<String> verifyUser(@Query("email") String email,
                                       @Query("token") String token,
                                       @Query("response_type") ResponseType responseType);

  /**
   * Resets the verification status of the user with the given email. This will set the verification
   * status of the user to false and reset the associated verification token.
   *
   * @param email the user's email address
   * @param password the user's password
   * @return a {@link CompletableFuture} that holds the updated user after the request completes
   */
  @POST("verify/reset")
  CompletableFuture<User> resetVerificationStatus(@Query("email") String email,
                                                  @Header("password") String password);

  /**
   * Get a new builder instance to build a {@code ThunderClient}.
   *
   * @return a new client builder
   */
  static ThunderClientBuilder builder() {
    return new ThunderClientBuilder();
  }

  /**
   * Create a new {@link ThunderClientFake} to be used for testing purposes. Note that the
   * password header will be required for requests with this method. To disable the password
   * header check, use {@link #fake(boolean)}.
   *
   * @return a new instance of {@link ThunderClientFake}
   */
  static ThunderClient fake() {
    return fake(true);
  }

  /**
   * Create a new {@link ThunderClientFake} to be used for testing purposes.
   *
   * @param requirePasswordHeader whether or not to require the password header to be present
   *                              when making requests
   * @return a new instance of {@link ThunderClientFake}
   */
  static ThunderClient fake(boolean requirePasswordHeader) {
    return new ThunderClientFake(requirePasswordHeader);
  }
}
